### ISSUE TITLE
How to configure keypoints_from_regression in tensorrt

### ISSUE THREAD

**Issue by zachluo** [open-mmlab/mmdeploy#386]
Hi, I wanna configure the post-processing keypoints_from_regression in converting a mmpose model into a tensorrt engine. How do I configure? Thanks!

**Comment by irexyc** [comment]
What configure do you want？

The keypoints_from_regression in DeepposeRegressionHead seems don't have any parameters to configure

**Comment by zachluo** [comment]
@irexyc The tensorrt engine for mmpose outputs the heatmap now. I wanna produce the keypoints with a keypoints_from_regression plugin.

**Comment by irexyc** [comment]
If you convert the model(in mmpose repo) with mmdeploy, the model should support  keypoints_from_regression. According to csrc\apis\c\pose_detector.h, the final output should be a n length vector contain point and score.

If so, which convert command do you used? And which api do you want to use, python or c?

**Comment by zachluo** [comment]
@irexyc I follow this instruction with python.
https://github.com/open-mmlab/mmdeploy/blob/master/docs/en/codebases/mmpose.md

Yes, I saw the declaration in csrc\apis\c\pose_detector.h. Could you provide a demo of how to configure keypoints_from_regression in config (configs/mmpose/posedetection_tensorrt_static-256x192.py)?

**Comment by irexyc** [comment]
This model config use regression postprocess, take this model for example [mmpose/configs/body/2d_kpt_sview_rgb_img/deeppose/coco/res50_coco_256x192.py]()

1) convert the model:
``` bash
python tools/deploy.py \
    $DEPLOY_CFG \
    $MODEL_CFG \
    $MODEL_PTH \
    $IMG \
    --device $DEVICE \
    --dump-info \
    --work-dir $MODEL 

python tools/deploy.py configs/mmpose/posedetection_tensorrt_static-256x192.py \
  mmpose/configs/body/2d_kpt_sview_rgb_img/deeppose/coco/res50_coco_256x192.py \
  mmpose/checkpoints/body/2d_kpt_sview_rgb_img/deeppose/coco/deeppose_res50_coco_256x192-f6de6c0e_20210205.pth
  demo\resources\human-pose.jpg \
  --device cuda \
  --work_dir work_dir \
  --dump-info
```
When you add --dump-info, it will generate a pipeline.jsjon which express the pipeline, you can see this part which describe postproces in pipeline.json. Currently, the sdk support 2d pose detection with keypoints_from_regression and keypoints_from_heatmap decode.
```
{
    "type": "Task",
    "module": "mmpose",
    "name": "postprocess",
    "component": "DeepposeRegressionHeadDecode",
    "params": {
        "flip_test": true
    },
    "output": [
        "post_output"
    ],
    "input": [
        "prep_output",
        "infer_output"
    ]
}
```

2) compile the c demo, and run the demo. The input parameter model_path is the previous work_dir.
```
./pose_detection cuda work_dir img_path
```

**Comment by zachluo** [comment]
@irexyc Thanks for your advice! But I am still confused about the configuration of keypoints_from_regression with the Tensorrt backen. Could you give more details?

**Comment by irexyc** [comment]
You don't have to configure it.

If the model is a "standard model"(from mmxx repo and not modified) that mmdeploy can convert,  when convert the model, the pre/post process information will automatically generated in pipeline.json, so the sdk framework can read the json and do the correspond process.

**Comment by zachluo** [comment]
@irexyc Then that is great! But I found that the output shape is still a heatmap (1, 17, 64, 48) when the tensorrt engine is applied to Triton Server. Should I modify the config by codebase_config = dict(model_type='sdk')?

**Comment by zachluo** [comment]
yes, you now get my point. I use python api to produce a tensorrt engine and use it to the triton which I need all the pre, forward, and post in one engine rather than you said the sdk read the pipeline and do the post. Thanks for your kind help again!



---Original---
From: "Chen ***@***.***&gt;
Date: Thu, Apr 21, 2022 21:12 PM
To: ***@***.***&gt;;
Cc: "Weixin Luo ***@***.******@***.***&gt;;
Subject: Re: [open-mmlab/mmdeploy] How to configure keypoints_from_regression in tensorrt (Issue #386)




 
I get what you meant now. You can treat the pipeline like (preprocess) -&gt; (model forward) -&gt; (postprocess). The direct trt model output(model forward output) is  a heatmap (1, 17, 64, 48), the sdk framework read the pipeline and do the postprocess which convert the heatmap to the final result.
 
If you use c api, you don't need the config, all you need is the work_dir.
 
I think what you want is python api? I find the current pose detector python api has some issues. I will fix it soon and give a feedback tomorrow.
 
Currently I want to make sure what you want is below, take object detection as example :
 from mmdeploy_python import Detector import cv2 detector = Detector("path/to/modeldir/", "cuda", 0) img = cv2.imread("path/to/image")  result = detector.Apply(img)

**Comment by irexyc** [comment]
@zachluo Sorry for late replay. The fix(#396) is under review, you can test that branch first.

**Comment by zachluo** [comment]
> @zachluo Sorry for late replay. The fix(#396) is under review, you can test that branch first.
> 
> ```python
> from mmdeploy_python import PoseDetector
> import cv2
> 
> handle = PoseDetector("E:/pjlab/testcs/trt", "cuda", 0)
> img = cv2.imread(r"E:\pjlab\dev-v0.4.0\demo\resources\human-pose.jpg")
> result = handle([img]) # take full image as a bbox
> 
> # OR
> # boxes = [[0,0,img.shape[1] - 1, img.shape[0] - 1]]
> # result = handle([img], [boxes])
> ```

@irexyc great! I am now trying it.

**Comment by zachluo** [comment]
Now the problem is solved and this issue is gonna be closed.
